inst/tests/froll.Rraw

```diff
@@ -433,32 +433,43 @@ ma = function(x, n, na.rm=FALSE, nf.rm=FALSE) {
   if (!is.double(x)) x = as.double(x)
   if (!is.integer(n)) n = as.integer(n)
   ans = rep(NA_real_, nx<-length(x))
-  if (nf.rm) x[!is.finite(x)] = NA_real_ # exact=F consistency due to https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17441
+  if (nf.rm) x[!is.finite(x)] = NA_real_ # algo=fast consistency due to https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17441
   for (i in n:nx) ans[i]=mean(x[(i-n+1):i], na.rm=na.rm)
   ans
 }
+compiler_o_flag_match = function() {
+  ## escape tests for NA/NaN exactness if compiled with -O0 flag so NaN behaviour changes: #3353 needed only for .123 and .162 but few other tests might be affected on different platforms also
+  ## related to https://bugs.r-project.org/bugzilla/show_bug.cgi?id=17441
+  x = c(NA, NaN)
+  y = c(NaN, NA)
+  `&&`(
+    identical(mean(x), frollmean(x, 2L, algo="exact")[2L]),
```

Comment:
Better to test the `x` and `y` independently against fixed known results (`NA` or `NaN` in this case). Then I for example as reader of the test can more easily see what is happening here.  For example, I'm currently wondering if base R changes result depending on whether R itself has been compiled with -O0 or not, or whether it's just frollmean that changes result depending on whether data.table has been compiled with -O0 or not.
Also the wording that this behavior depends on -O0 seems odd.  It's really undefined or variable behaviour that -O0 picks up currently.  But equally it could happen on other platforms:  Solaris and PowerPC spring to mind.
How about changing the tests to test if either NA or NaN are returned?  The PR currently turns off the tests, so in the event that `Inf` was returned (a definite error), the test wouldn't run because it was turned off.  Better to always test that one of the accepted values is always returned.  `is.na(NA)` and `is.na(NaN)` both return `TRUE`, so `is.na()` could be used by the test.

Reply:
> Equally, couldn't it happen on other platforms? Solaris and Power spring to mind.

NaN - NA interaction is not undefined by any standards, thus is platform specific, and what we can observe here, also `-O` flag specific.

----

> How about changing the tests to test if _either_ NA or NaN are returned?

Those particular tests are there to test `algo="exact"` which has to be identical to base R result. [`froll` manual](https://rdatatable.gitlab.io/data.table/library/data.table/html/froll.html) states:

> When set to `"exact"` ... carefully handles all non-finite values.

and

> It also handles NaN, +Inf, -Inf consistently to base R.

There are other tests for `algo="fast"` where NaN as treated as NA.
Agree on that it is better to ignore NaN-NA difference than just disabling those tests. I will amend.
BTW. slightly related #2960

----

> Better to always test that one of the accepted values is always returned.

Best to not "always" but only when issue that issue occurs.